native/java/org/jpype/proxy/JPypeProxy.java

```diff
@@ -69,35 +71,46 @@ public Object newInstance()
   public Object invoke(Object proxy, Method method, Object[] args)
           throws Throwable
   {
-    try
-    {
-//      context.incrementProxy();
-      if (context.isShutdown())
-        throw new RuntimeException("Proxy called during shutdown");
 
-      // We can save a lot of effort on the C++ side by doing all the
-      // type lookup work here.
-      TypeManager typeManager = context.getTypeManager();
-      long returnType;
-      long[] parameterTypes;
-      synchronized (typeManager)
+    if (context.isShutdown())
+      throw new RuntimeException("Proxy called during shutdown");
+
+    // We can save a lot of effort on the C++ side by doing all the
+    // type lookup work here.
+    TypeManager typeManager = context.getTypeManager();
+    long returnType;
+    long[] parameterTypes;
+    synchronized (typeManager)
+    {
+      returnType = typeManager.findClass(method.getReturnType());
+      Class<?>[] types = method.getParameterTypes();
+      parameterTypes = new long[types.length];
+      for (int i = 0; i < types.length; ++i)
       {
-        returnType = typeManager.findClass(method.getReturnType());
-        Class<?>[] types = method.getParameterTypes();
-        parameterTypes = new long[types.length];
-        for (int i = 0; i < types.length; ++i)
-        {
-          parameterTypes[i] = typeManager.findClass(types[i]);
-        }
+        parameterTypes[i] = typeManager.findClass(types[i]);
       }
+    }
 
-      return hostInvoke(context.getContext(), method.getName(), instance, returnType, parameterTypes, args);
-    } finally
+    // Check first to see if Python has implementated it
+    Object result = hostInvoke(context.getContext(), method.getName(), instance, returnType, parameterTypes, args, missing);
+    
+    // If we get a good result than return it
+    if (result != missing)
+      return result;
+    
+    // If it is a default method in the interface then we have to invoke it using special reflection.
+    if (method.isDefault())
     {
-//      context.decrementProxy();
+      return MethodHandles.lookup()
+              .unreflectSpecial(method, method.getDeclaringClass())
+              .bindTo(proxy)
+              .invokeWithArguments(args);
```

Comment:
Wouldn't using [Method handles.publicLookup](https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/MethodHandles.html#publicLookup--) be better here since it can only access public fields/methods thus removing the privileged issue?

I'm sure there are some possible benefits to having a java agent, I just wanted to point this out.

Reply:
We can certainly try that.   Though we will still need the agent if we want to make sure that it never changes behavior based on how JPype is started.   (Something that plagues the development.)

Reply:
The other reason for requiring privilege is if we ever want to be able to "extend" a class in Python.   In that case we will need to get to protected methods, and protected fields.

Reply:
> The other reason for requiring privilege is if we ever want to be able to "extend" a class in Python.   In that case we will need to get to protected methods, and protected fields.

I actually have a prototype using JDK22 ClassFile API preview feature, as an independent python package, where I managed this. I designed it around the impression that if you want to extend a Java class from Python, then you don't care about performance or bloat.

I basically created delegate methods that start with "-" which could only be accessed by reflection or it would be a syntax error. It's gross but it worked.

I was considering integrating and contributing it since it would not introduce GPL code. Was just waiting on it to exit preview in a LTS JDK and then for the free time.

Reply:
You should be able to call a default method from jni using `CallNonvirtualObjectMethodA` if you use the interface with the default method as the claz parameter.

I figured this out while working on `super()` support for extension classes.

Reply:
CallNonvirtualObjectMethodA does not work.  I ran tests on it.   While it does allow the default method to be called, it ALWAYS calls the default even when overwritten.   I thus went with this approach as there are no methods in the JNI interface that gave proper behavior.

Reply:
> CallNonvirtualObjectMethodA does not work.  I ran tests on it.   While it does allow the default method to be called, it ALWAYS calls the default even when overwritten.   I thus went with this approach as there are no methods in the JNI interface that gave proper behavior.

Yes that is what the NonVirtual does, it's invoke exactly the specified method. Thinking about it, I'm not sure why the regular CallVirtual doesn't work.

Reply:
The reason it doesn't is that for interfaces "ALL" virtual method route to the proxy dispatch.   As default methods are virtual that means they effectively are not reachable other than by the "NonVirtual" method.   We could in principle your the virtual call request to our method, then call to Python with a reference handle find out that it was really supposed to go to default, then reroute back in the JNI method for a NonVirtual call.   That would mean unpacking and repacking arguments at C level and keeping track of the default level in C.  Of course we would have to know to try the nonvirtual path (trying to call it when it doesn't exist is a segfault).   So that means everything needs to be handed to the the C method.   I attempted that path but found it distasteful as it required making reflection API calls prior to launching to C.

It is "doable" but means more routing on the primary path to check for and passing the default method down to C.   That would make this already slow path potentially even slower for a feature that not many people will likely use.   Better to reroute back to Java and try to invoke the handle there.